File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.Stack;

class Solution {
    public int solution(String s) {
        if (s.length() % 2 != 0) return 0;
        
        int answer = 0;

        for (int x = 0; x < s.length(); x++) {
            String rotated = s.substring(x) + s.substring(0, x);
            if (isValid(rotated)) {
                answer++;
            }
        }
        return answer;
    }

    private boolean isValid(String str) {
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);

            if (ch == '(' || ch == '[' || ch == '{') {
                stack.push(ch);
            } else {
                if (stack.isEmpty()) return false;

                char top = stack.peek();
                if ((ch == ')' && top == '(') ||
                    (ch == ']' && top == '[') ||
                    (ch == '}' && top == '{')) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    if len(s) % 2 != 0:
        return 0

    answer = 0
    n = len(s)

    for x in range(n):
        rotated = s[x:] + s[:x]

        if is_valid(rotated):
            answer += 1

    return answer


def is_valid(str_val):
    stack = []
    pair = {")": "(", "]": "[", "}": "{"}

    for ch in str_val:
        if ch in "([{":
            stack.append(ch)
        else:
            if not stack or stack[-1] != pair[ch]:
                return False
            stack.pop()

    return len(stack) == 0